### PR TITLE
Undeclared data categories

### PR DESCRIPTION
**Pull request by galvana**
Closes [PROD-1873](https://ethyca.atlassian.net/browse/PROD-1873)

### Description Of Changes

Updating datamap report table with two new columns:

- `System undeclared data categories` - A list of data categories aggregated from all of a system's datasets that are not declared in any of the system's data uses
- `Data use undeclared data categories` - A list of data categories from a data use's dataset that are not declared in the current data use or any sibling data uses

### Code Changes

* [ ] Adding new columns to `DatamapReportTable.tsx`
* [ ] Misc fixes for front-end cache invalidation
* [ ] Helper functions on `System` and `PrivacyDeclaration` models to aggregate and report on associated data categories

### Steps to Confirm

* [x] Start fidesplus from the branch in (https://github.com/ethyca/fidesplus/pull/1384)
* [x] Start the Admin UI in fides
* [x] Navigate to Data Inventory > Manage Datasets
* [x] Upload the [Postgres sample dataset](https://raw.githubusercontent.com/ethyca/fides/main/data/dataset/postgres_example_test_dataset.yml)

**System undeclared data categories**
* [x] Navigate to Data Inventory > Add System, set the **System name** and set **Dataset references**
 to the Postgres sample dataset, click **Save**
* [x] Navigate to Data Inventory > Reporting, the `System undeclared data categories` should not be visible
* [x] Click on **Edit columns** and enable the `System undeclared data categories` column, click **Save**
* [x] The enabled column should be visible and expandable
* [x] `System undeclared data categories` should say "12 system undeclared data categories"
* [x] Navigate back to the system and add a new data use with the `user` data category
* [x] Navigate back to the datamap report table. Enable the column again. The `System undeclared data categories` column should now show a single undeclared data category, `System Data: Operations Data`

**Data use undeclared data categories**
* [x] Remove the dataset from the system, create a new data use on the system, set the data category to `System` and add the Postgres dataset
* [x] Navigate to Data Inventory > Reporting, the `Data use undeclared data categories` should not be visible
* [x] Click on **Edit columns** and enable the `Data use undeclared data categories` column, click **Save**
* [x] The enabled column should be visible and expandable
* [ ] `Data use undeclared data categories` should say "11 system undeclared data categories"
* [x] Navigate back to the system and change the data category on the data use to `user`
* [x] Navigate back to the datamap report table. Enable the column again. The `Data use undeclared data categories` column should now show a single undeclared data category, `System Data: Operations Data`

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* [ ] Issue Requirements are Met
* [ ] Update `CHANGELOG.md`

[PROD-1873]: https://ethyca.atlassian.net/browse/PROD-1873?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: clients/admin-ui/src/features/common/table/v2/column-settings/ColumnSettingsModal.tsx [hunk 1]
```diff
@@ -47,7 +47,8 @@ export const ColumnSettingsModal = <T,>({
         .map((c) => ({
           id: c.id,
           displayText: c.columnDef?.meta?.displayText || c.id,
-          isVisible: c.getIsVisible(),
+          isVisible:
+            tableInstance.getState().columnVisibility[c.id] ?? c.getIsVisible(),
```
Comment (galvana) [inline clients/admin-ui/src/features/common/table/v2/column-settings/ColumnSettingsModal.tsx:51]: This is to allow us to set default visibility, see: https://github.com/ethyca/fides/pull/4781/files#diff-cd47c40887e69a55fee3a54b819214c466bf33045c6be67473345c87cef25bbcR983-R986

File: clients/admin-ui/src/features/datamap/datamap.slice.ts [hunk 2]
```diff
@@ -102,6 +102,7 @@ const datamapApi = baseApi.injectEndpoints({
           url: `plus/datamap/minimal?${queryString}`,
         };
       },
+      providesTags: ["Datamap"],
```
Comment (galvana) [inline clients/admin-ui/src/features/datamap/datamap.slice.ts:105]: This is to allow the datamap report table to reload when there are changes to systems

File: clients/admin-ui/src/features/datamap/modals/SettingsModal.tsx [hunk 3]
```diff
@@ -271,7 +271,7 @@ const SettingsModal: React.FC<SettingsModalProps> = ({ isOpen, onClose }) => {
     <Modal isOpen={isOpen} onClose={onClose} isCentered size="2xl">
       <ModalOverlay />
       <ModalContent>
-        <ModalHeader pb={0}>Data Map Settings</ModalHeader>
+        <ModalHeader pb={0}>Data map settings</ModalHeader>
```
Comment (galvana) [inline clients/admin-ui/src/features/datamap/modals/SettingsModal.tsx:274]: Changing to sentence case

File: clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx [hunk 4]
```diff
@@ -866,6 +868,62 @@ export const DatamapReportTable = () => {
...
+          <DefaultHeaderCell
+            value="Data use undeclared data categories"
+            {...props}
+          />
+        ),
+        meta: {
+          displayText: "Data use undeclared data categories",
+          showHeaderMenu: true,
+        },
+      }),
```
Comment (galvana) [inline clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx:926]: New column definitions

File: clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx [hunk 5]
```diff
@@ -922,6 +980,10 @@ export const DatamapReportTable = () => {
     data,
     initialState: {
       columnOrder,
+      columnVisibility: {
+        [COLUMN_IDS.SYSTEM_UNDECLARED_DATA_CATEGORIES]: false,
+        [COLUMN_IDS.DATA_USE_UNDECLARED_DATA_CATEGORIES]: false,
+      },
```
Comment (galvana) [inline clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx:986]: Used to disable the undeclared data columns by default

File: clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx [hunk 6]
```diff
@@ -1023,7 +1085,7 @@ export const DatamapReportTable = () => {
             </MenuList>
           </Menu>
           <Button
-            data-testid="filter-multiple-systems-btn"
+            data-testid="edit-columns-btn"
```
Comment (galvana) [inline clients/admin-ui/src/features/datamap/reporting/DatamapReportTable.tsx:1088]: The two buttons in this group had duplicate `data-testid`s

File: requirements.txt [hunk 7]
```diff
@@ -38,6 +38,7 @@ pyinstrument==4.5.1
 psycopg2-binary==2.9.6
 pydantic==1.10.9
 pydash==6.0.2
+pygtrie==2.5.0
```
Comment (galvana) [inline requirements.txt:41]: Open-source trie library used for the data category matching

File: src/fides/api/models/sql_models.py [hunk 8]
```diff
@@ -450,10 +507,19 @@ class PrivacyDeclaration(Base):
         nullable=False,
         index=True,
     )
-    system = relationship(System, back_populates="privacy_declarations")
+    system = relationship(
+        System, back_populates="privacy_declarations", lazy="selectin"
+    )
```
Comment (galvana) [inline src/fides/api/models/sql_models.py:512]: Need to be careful of this change, it makes the system retrieval on privacy declarations eager

**Comment by pattisdr** [comment]
Plan to start review Wed 4/17 -

File: src/fides/api/util/taxonomy_utils.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    This function uses a trie-based approach to perform hierarchical matching of categories. It considers
+    a category as declared if it matches any of the following conditions:
+    - The category itself is present in the declared categories.
+    - Any parent category of the category is present in the declared categories.
+
+    For example, if "user.contact.email" is in input_categories and the declared categories contain
+    "user.contact" or "user", then "user.contact.email" will be considered as declared.
+    """
+
+    declared_trie = StringTrie(separator=".")
```
Comment (pattisdr) [inline src/fides/api/util/taxonomy_utils.py:21]: whoa fancy 👀

File: src/fides/api/models/sql_models.py [hunk 10]
```diff
@@ -191,6 +193,22 @@ def from_fideslang_obj(
...
+
+        Example inputs and outputs:
+            - `a.b.c` --> [`a.b.c`, `a.b`, `a`]
+            - `a` --> [`a`]
+        """
+        parent_categories = {data_category_key}
+        while data_category_key := data_category_key.rpartition(".")[0]:
+            parent_categories.add(data_category_key)
+        return parent_categories
+
```
Comment (pattisdr) [inline src/fides/api/models/sql_models.py:211]: This looks familiar

EDIT: Ah yes! data uses. we could consider drying it up and sharing the method -

File: src/fides/api/models/sql_models.py [hunk 11]
```diff
@@ -493,6 +559,27 @@ def purpose(cls) -> Case:
...
+        """
+
+        # all data categories from the datasets
+        dataset_data_categories = set()
+        for dataset in self.datasets:
+            dataset_data_categories.update(dataset.field_data_categories)
+
+        # all data categories specified directly on this and sibling privacy declarations
+        declared_data_categories = set()
+        for privacy_declaration in self.system.privacy_declarations:
```
Comment (pattisdr) [inline src/fides/api/models/sql_models.py:576]: I'm a little confused about the organization here.  So each privacy declaration has undeclared data categories that are looking at itself but also all of the other privacy declarations? Should this property be at this level? (instead a property on the system, maybe, privacy_declaration_undeclared_data_categories?

**Review by pattisdr — COMMENTED** [review]
this is looking great, I just have a few questions, also these are the best testing instructions I've ever seen. Really helpful for someone that wasn't involved in the conversations for this feature.

File: src/fides/api/models/sql_models.py (hunk 11 above)
Comment (galvana) [reply src/fides/api/models/sql_models.py:576]: Ah yes, there's a tiny detail here which is that the `undeclared_data_categories` on the `PrivacyDeclaration` report on the data categories tied to the `datasets` on the `PrivacyDeclaration`. So the search space is all of the data categories across all PDs but for each PD we're only evaluating the datasets associated with itself.

Comment (pattisdr) [reply src/fides/api/models/sql_models.py:576]: ah ok can you make this a code comment?

File: src/fides/api/models/sql_models.py (hunk 8 above)
Comment (pattisdr) [reply src/fides/api/models/sql_models.py:512]: Are you concerned about performance on this PR?

File: src/fides/api/models/sql_models.py (hunk 10 above)
Comment (galvana) [reply src/fides/api/models/sql_models.py:211]: I'll remove it for now, I didn't end up needing it

**Review by pattisdr — APPROVED** [review]

File: src/fides/api/models/sql_models.py (hunk 8 above)
Comment (pattisdr) [reply src/fides/api/models/sql_models.py:496]: Systems already feel very bogged down, at some point we may reach a tipping point where we need to rearchitect some of this -

**Pull request merged**